inductiva/molecules/scenarios/protein_solvation/protein_solvation.py

```diff
@@ -86,28 +105,67 @@ def simulate_async(
         Args:
             simulation_time: The simulation time in ns.
             integrator: The integrator to use for the simulation. Options:
-                - "md" (Molecular Dynamics): Accurate leap-frog algorithm for 
+                - "md" (Molecular Dynamics): Accurate leap-frog algorithm for
                 integrating Newton's equations of motion.
                 - "sd" (Steepest Descent): Stochastic dynamics integrator with
                 leap-frog scheme.
-                - "bd" (Brownian Dynamics): Euler integrator for Brownian or 
-                position Langevin dynamics. 
-                
-            For more details on the integrators, refer to the GROMACS 
-            documentation at 
+                - "bd" (Brownian Dynamics): Euler integrator for Brownian or
+                position Langevin dynamics.
+
+            For more details on the integrators, refer to the GROMACS
+            documentation at
             https://manual.gromacs.org/current/user-guide/mdp-options.html.
 
             nsteps_minim: Number of steps for energy minimization.
         """
+
+        #Compute charge if it is not provided
+        if self.charged is None:
+            self.charged = self.compute_charge()
+
+        #Edit commands.json according to the charge of the protein
+        commands_path = os.path.join(self.template_dir, "commands.json")
+        replace_params_in_template(self.template_dir, "commands.json.jinja", {
+            "warnings": "-maxwarn 1",
+            "solvents": "SOL"
+        }, commands_path)
+
         self.nsteps = int(
             simulation_time * 1e6 / 2
         )  # convert to fs and divide by the time step of the simulation (2 fs)
         self.integrator = integrator
         self.nsteps_minim = nsteps_minim
-        commands = self.read_commands_from_file(
-            os.path.join(self.template_dir, "commands.json"))
+
+        commands = self.read_commands_from_file(commands_path)
         return super().simulate_async(simulator, commands=commands)
 
+    def compute_charge(self, simulator: Simulator = GROMACS()):
+        """Check if the protein is charged."""
+
+        logging.info("Computing the charge of the protein")
+
+        output_dir = os.path.dirname(self.protein_pdb)
```

Comment:
I still vouch to execute the command below only once, instead of repeating it in the `commands.json.jinja`...

Hence, here you could immediately write to your current folder, instead of creating a new one just for this purpose. Otherwise, you could use a temporary folder, that disappears after this execution.